### ISSUE TITLE
Incorrect warning about hidden overloaded virtual function with virtual inheritance

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#11936]
|  |  |
| --- | --- |
| Bugzilla Link | [11564](https://llvm.org/bz11564) |
| Version | trunk |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @tkremenek |

## Extended Description 
The following code generates a warning in clang 3.0:

```c++
struct A {
    virtual ~A() {}
    virtual void func(int) = 0;
};

struct B {
    virtual ~B() {}
    virtual void func(double) = 0;
};

struct Base: public virtual A {
    virtual void func(int) {}
};

struct Derived: public Base, virtual public A, virtual public B {
    virtual void func(int) {}
    virtual void func(double) {}
};

int main() {
    Derived d;
    d.func(1);
    d.func(1.0);
}
```

./clang++ -Wall -o test test.cpp
```c++
test.cpp:17:18: warning: 'Derived::func' hides overloaded virtual function [-Woverloaded-virtual]
    virtual void func(double) {}
                 ^
test.cpp:3:18: note: hidden overloaded virtual function 'A::func' declared here
    virtual void func(int) = 0;
                 ^
1 warning generated.
```

Note that func(int) is implemented in Derived, so it is not hidden and the warning is incorrect. gcc 4.6.2 processes this example correctly and does not generate any warnings.

**Comment by philnik777** [comment]
Fixed in Clang 3.2: https://godbolt.org/z/vYzT67rjv

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-c-1
